certbot-apache/certbot_apache/augeasparser.py

```diff
@@ -0,0 +1,192 @@
+""" Tests for ParserNode interface """
+from certbot_apache import interfaces
+
+from acme.magic_typing import Dict, Tuple  # pylint: disable=unused-import, no-name-in-module
+
+
+class AugeasCommentNode(interfaces.CommentNode):
+    """ Augeas implementation of CommentNode interface """
+    ancestor = None
+    comment = ""
+    dirty = False
+    _metadata = dict()  # type: Dict[str, object]
+
+    def __init__(self, comment, ancestor=None):
+        self.comment = comment
+        self.ancestor = ancestor
+
+    def save(self, msg):  # pragma: no cover
+        pass
+
+    # Apache specific functionality
+
+    def get_metadata(self, key):
+        """ Returns a metadata object
+
+        :param str key: Metadata object name to return
+        :returns: Requested metadata object
+        """
+        try:
+            return self._metadata[key]
+        except KeyError:
+            return None
+
+class AugeasDirectiveNode(interfaces.DirectiveNode):
+    """ Augeas implementation of DirectiveNode interface """
+    ancestor = None
+    parameters = tuple()  # type: Tuple[str, ...]
+    dirty = False
+    enabled = True
+    name = ""
+    _metadata = dict()  # type: Dict[str, object]
+
+    def __init__(self, name, parameters=tuple(), ancestor=None):
+        self.name = name
+        self.parameters = parameters
+        self.ancestor = ancestor
+
+    def save(self, msg):  # pragma: no cover
+        pass
+
+    def set_parameters(self, parameters):  # pragma: no cover
+        self.parameters = tuple("CERTBOT_PASS_ASSERT")
+
+    # Apache specific functionality
```

Comment:
This sounds good. I'm doing this change in the same PR as the change from properties to PEP 526 style definition.

Reply:
What do you think about doing it in this PR?

My thought is that by doing it here, any concerns about the details of those methods block one PR rather than two.

Reply:
Oh, I missed this comment and implemented it in the same, smaller PR, that this one depends on anyway. Should I move it here instead? I think I'd personally prefer keeping them in the smaller one, as I think it's easier and faster to review and merge.

The `has_parameter()` utility function wasn't implemented there, as I think it's more suitable for the utility mixin.

Reply:
> Should I move it here instead? I think I'd personally prefer keeping them in the smaller one, as I think it's easier and faster to review and merge.

I don't care that much but I still have my original preference. I added a number of comments to this PR suggesting changes to those methods and I think it'd be nice to keep that conversation in one place. I also think quickly landing the changes to how attributes are documented unblocks more work such as the majority of the combined `ParserNode` object.

> The has_parameter() utility function wasn't defined there, as I think it's more suitable for the utility mixin.

I think the implementation being a mixin makes sense, but if a ParserNode object is required to implement `has_parameter` to be used by the configurator, I think it should be documented in the interface as well.